source/windows/bcrypt_aes.c

```diff
@@ -0,0 +1,1110 @@
+
+    cipher->cipher.impl = cipher;
+    cipher->cipher.good = true;
+
+    return &cipher->cipher;
+
+error:
+    if (cipher != NULL) {
+        s_aes_default_destroy(&cipher->cipher);
+    }
+
+    return NULL;
+}
+
+/* This is just an encrypted key. Append them to a buffer and on finalize export/import the key using AES keywrap. */
+static int s_key_wrap_encrypt_decrypt(
+    struct aws_symmetric_cipher *cipher,
+    const struct aws_byte_cursor input,
+    struct aws_byte_buf *out) {
+    (void)out;
+    struct aes_bcrypt_cipher *cipher_impl = cipher->impl;
+
+    return aws_byte_buf_append_dynamic(&cipher_impl->overflow, &input);
+}
+
+/* Import the buffer we've been appending to as an AES key. Then export it using AES Keywrap format. */
+static int s_keywrap_finalize_encryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out) {
+    struct aes_bcrypt_cipher *cipher_impl = cipher->impl;
+
+    BCRYPT_KEY_HANDLE key_handle_to_encrypt =
+        s_import_key_blob(s_aes_keywrap_algorithm_handle, cipher->allocator, &cipher_impl->overflow);
+
+    NTSTATUS status = 0;
+
+    ULONG output_size = 0;
+    /* Call with NULL first to get the required size. */
+    status = BCryptExportKey(
+        key_handle_to_encrypt, cipher_impl->key_handle, BCRYPT_AES_WRAP_KEY_BLOB, NULL, 0, &output_size, 0);
+
+    if (!NT_SUCCESS(status)) {
+        cipher->good = false;
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+
+    int ret_val = AWS_OP_ERR;
+
+    if (aws_symmetric_cipher_try_ensure_sufficient_buffer_space(out, output_size)) {
+        goto clean_up;
+    }
+
+    /* now actually export the key */
+    ULONG len_written = 0;
+    status = BCryptExportKey(
+        key_handle_to_encrypt,
+        cipher_impl->key_handle,
+        BCRYPT_AES_WRAP_KEY_BLOB,
+        out->buffer + out->len,
+        output_size,
+        &len_written,
+        0);
+
+    if (!NT_SUCCESS(status)) {
+        cipher->good = false;
+        goto clean_up;
+    }
+
+    out->len += len_written;
+
+    ret_val = AWS_OP_SUCCESS;
+
+clean_up:
+    if (key_handle_to_encrypt) {
+        BCryptDestroyKey(key_handle_to_encrypt);
+    }
+
+    return ret_val;
+}
+
+/* Import the buffer we've been appending to as an AES Key Wrapped key. Then export the raw AES key. */
+
+static int s_keywrap_finalize_decryption(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out) {
+    struct aes_bcrypt_cipher *cipher_impl = cipher->impl;
+
+    BCRYPT_KEY_HANDLE import_key = NULL;
+
+    /* use the cipher key to import the buffer as an AES keywrapped key. */
+    NTSTATUS status = BCryptImportKey(
+        s_aes_keywrap_algorithm_handle,
+        cipher_impl->key_handle,
+        BCRYPT_AES_WRAP_KEY_BLOB,
+        &import_key,
+        NULL,
+        0,
+        cipher_impl->overflow.buffer,
+        (ULONG)cipher_impl->overflow.len,
+        0);
+    (void)status;
+
+    int ret_val = AWS_OP_ERR;
+
+    if (import_key) {
```

Comment:
if `BCryptImportKey()` fails, and import_key is NULL, we never actually aws_raise_error() before we return AWS_OP_ERR